bigsi/__main__.py

```diff
@@ -130,6 +126,88 @@ def bloom(self, ctx, outfile, config=None):
             kmers=extract_kmers_from_ctx(ctx, config["k"]),
         )
 
+    @hug.object.cli
+    def merge_blooms(
+        self,
+        from_file: hug.types.text = None,
+        out_file: hug.types.text = None,
+        num_rows: hug.types.number = None
```

Comment:
`num_rows` is an attribute of bloom filters/bigsi indexes that is used in many BIGSI commands. It looks like the convention in BIGSI is to get this property through a `config` file, instead of asking for it as a parameter. `merge_blooms` is also now the only CLI command that does not require a `config` file. I wonder if it would be better to require a `config` file here to keep this command  consistent with the others

Reply:
I thought about this when implementing this subcommand. In my opinion, the subcommands `merge_blooms` and `bloom` are about creating a bloom filter array or matrix. That's not quite in the stage of building the index (i.e. create the db, inserting the db etc.) or querying the index.

The config file typically contains a lot of other settings that are not needed for `bloom` and `merge_bloom`. For `bloom`, only the number of rows and the number of hash functions are required. For `merge_blooms`, only the number of rows and the number of columns (provided through `from_file` argument) are required as it simply merge multiple matrices.

Reply:
I totally agree. I would prefer to change the `bloom` and `merge_bloom` commands to not receive `config` files, and receive only the information it needs to build bloom filters (adhering to the principle of least privilege).

I would like also to make bloom filters self-contained data structures. Right now, if we have a bloom filter file, we don't know exactly how many rows it has (we don't know how many bits were used in the last byte for padding), how many hash functions it used to set bits, nor how many samples and which samples. We always need metadata associated to a bloom filter, and this is error prone. I think bloom filters should be self-contained, like indexes.

If you agree with this, I think we should raise issues to solve this, but not actually address this in this PR

Reply:
I agree with you on this. I was thinking using some custom headers to maintain the metadata for the bitmatrix, or using some file formats like hdf5. But then I wanted to make it compatible with the current file format so that I can re-use all the blooms already created.

Reply:
Argh, this will indeed break backwards compatibility... unless we provide a script to transform a "raw" bloom filter into this new bloom filter format.

Reply:
A script may not be needed. We can design and work with two different file formats: self-contained one and headless one. Try the self-contained one first and check if the file size makes sense. If not, treat it as headless one (need row and col input). If not again, panic!

Reply:
I would prefer the script as it would force the users to switch to the new bloom filter file format, and thus we pass the burden of ensuring compatibility to the users... or we take the burden of maintaining two different bloom file formats ourselves. Not a strong preference though, so I think you can make the decision!

Reply:
I prefer script too and let the users to fix their blooms. If we provide the script it is easy and quick to convert to new version. So we don't need to maintain two formats.